R/check_tbl_spl_n.R

```diff
@@ -0,0 +1,124 @@
+#' Check model output data tbl samples contain the appropriate number of samples
+#' for a given compound idx.
+#' @param tbl a tibble/data.frame of the contents of the file being validated. Column types must **all be character**.
+#' @inherit check_tbl_colnames params
+#' @inherit check_tbl_colnames return
+#' @export
+check_tbl_spl_n <- function(tbl, round_id, file_path, hub_path) {
+  config_tasks <- hubUtils::read_config(hub_path, "tasks")
+
+  if (isFALSE(has_spls_tbl(tbl)) || isFALSE(hubUtils::is_v3_config(config_tasks))) {
+    return(skip_v3_spl_check(file_path))
+  }
+
+  hash_tbl <- spl_hash_tbl(tbl, round_id, config_tasks)
+  n_ranges <- get_round_spl_n_ranges(config_tasks, round_id)
+
+  n_tbl <- dplyr::group_by(hash_tbl, .data$compound_idx) %>%
+    dplyr::summarise(
+      n = dplyr::n_distinct(.data$output_type_id),
+      mt_id = unique(.data$mt_id)
+    ) %>%
+    dplyr::left_join(n_ranges, by = "mt_id") %>%
+    dplyr::mutate(
+      less = .data$n < .data$n_min,
+      more = .data$n > .data$n_max,
+      out_range = .data$less | .data$more
+    ) %>%
+    dplyr::filter(.data$out_range)
```

Comment:
Right. Compound_idx is what defines a sample and we need to count samples across compound_idx to ensure they are within the correct range. See for eaxample how samples (output_type_ids) are marked up under different coumpound task id sets in the docs: https://hubverse.io/en/latest/user-guide/sample-output-type.html#four-submissions-differing-by-compound-modeling-task

Reply:
I've also opened the following issue. You can currently get this infomration from the hubData::expand_model_out_val_grid(include_sample_ids = TRUE)` but I think a function that creates a clear table illustrating compound idx structure of a given round/modeling task would be really useful https://github.com/Infectious-Disease-Modeling-Hubs/hubData/issues/40

Reply:
The part that I don't understand is why is it link to the output type id information? For me it's two different information. We are checking the output type id column in the other 2 functions with the compound ID checking, so if you have an error of ID, you will have an error but it does not mean you are missing samples. 
For example if someone provided the correct number of samples but made an error in the output type id column, than it should not return an error here but only about the output type id numbering.

Reply:
In the validation instriuctions it states:
> The number of rows for each combination of individual modeling task should fall between min_samples_per_task and max_samples_per_task (inclusive).

Compound_idx defines the individual modeling task of the combination unique compound task id values, so samples need to match a modeling task first to then be counted against the required number for that compound_idx.

I you have provided the wrong output type ID, that should cause an error here too because you have not properly specified the samples, so it would be wrong in view to count them as valid samples.

From the validation spec sheet, this the test I've tried to encode here:
> - Example: task ids are location, origin_date, horizon; compound units defined by combinations of location and origin_date; expect the number of samples for each such compound unit to be between specified min and max.
> - Logic: Within each group defined by a unique combination of values for the compound_taskid_set, the number of unique values for output_type_id should be between min_samples_per_task and max_samples_per_task

Reply:
I am sorry I miss and/or misunderstood some of the part in the instructions for the validation tests, for me the issue here in the "combination of individual modeling task" is are we talking about the minimum compound set the hub is asking or the compound set chosen by the team who made the submission?

Reply:
Don't worry! It is confusing...I'm confused  !

My interpretation is that combination of individual modeling tasks is defined by the `compund_taskid_set` in the config.  Each unique combination of values in the variables in the `compund_taskid_set` corresponds to a single `compound_idx`. And number of samples (i.e. unique `output_type_ids`) is counted separately for each `compound_idx`.

I note as well that some of this confusion is probably coming from the fact that in the description we do mention coarser samples passing validation (as you noted below) but I haven't implemented that as it wasn't in the validation instructions.

See below for my suggestion of how to proceed with this for now.

Reply:
Interestingly, I tested the validation with a coarser sample settings and it passes validation without any issue (JHU subset file) if the same compound id set is used everywhere. 
My hypothesis (but I am very not sure), is that it's because in the `get_mt_spl_hash_tb()` and in particular in the 
```r
split(tbl, f = tbl$output_type_id) %>%
    purrr::map(
      function(.x, compound_taskids, non_compound_taskids) {
        tibble::tibble(
          compound_idx = names(sort(table(.x$compound_idx), decreasing = TRUE))[1L],
          output_type_id = unique(.x$output_type_id),
          hash_comp_tid = rlang::hash(unique(.x[, compound_taskids])),
          hash_non_comp_tid = rlang::hash(.x[, non_compound_taskids]),
          hash_spl_id = rlang::hash(.x)
        )
      },
      non_compound_taskids = non_compound_taskids,
      compound_taskids = compound_taskids
    ) %>%
    purrr::list_rbind()
```
If all the samples share the same first compound id then it passes without issue.

Reply:
> I you have provided the wrong output type ID, that should cause an error here too because you have not properly specified the samples, so it would be wrong in my view to count them as valid samples against the expected number.

Also, I understand the logic behind the error but I also wonder if it's confusing for the user. 
For example, I still don't understand the output I got from my test. I have the expected number of sample, but hey are indeed misidentified. However, as they fit the minimal requirement of 100 samples per the hub compound id set test, I have no idea how to fix it. As we say before the error message might also need more information because for example I got and I don't know what the compound idx 10 means here as it's associated with different output type id:
```
Required samples per compound idx task not present.
  File contains less ("69") than the minimum required number of samples per task ("100") for
  compound idx "10" [...]
```

Reply:
If you dig into the `errors` object, it gives you a lot more detail including what the composition of each `compound_idx` and how many samples of that idx were successfully counted, eg. here's the first element of the output:
```r
$compound_idx
[1] "10"

$n
[1] 69

$min_samples_per_task
[1] 100

$min_samples_per_task
[1] 300

$compound_idx_tbl
# A tibble: 39,312 × 6
   origin_date scenario_id  location target   horizon age_group
   <chr>       <chr>        <chr>    <chr>    <chr>   <chr>    
 1 2024-04-28  A-2024-03-01 51       inc hosp 1       0-64     
 2 2024-04-28  A-2024-03-01 51       inc hosp 2       0-64     
 3 2024-04-28  A-2024-03-01 51       inc hosp 3       0-64     
 4 2024-04-28  A-2024-03-01 51       inc hosp 4       0-64     
 5 2024-04-28  A-2024-03-01 51       inc hosp 5       0-64     
 6 2024-04-28  A-2024-03-01 51       inc hosp 6       0-64     
 7 2024-04-28  A-2024-03-01 51       inc hosp 7       0-64     
 8 2024-04-28  A-2024-03-01 51       inc hosp 8       0-64     
 9 2024-04-28  A-2024-03-01 51       inc hosp 9       0-64     
10 2024-04-28  A-2024-03-01 51       inc hosp 10      0-64     
# ℹ 39,302 more rows
# ℹ Use `print(n = ...)` to see more rows
```

There's only so much detail I can add to the message (in fact looking at it it's too long atm!). Also after I've implemented https://github.com/Infectious-Disease-Modeling-Hubs/hubData/issues/40, users will have another eay to inspect expected hub compound idxs

Reply:
So sorry I was not clear, I dig into the errors message but still cannot find how to replicate the n=69, I know the solution here is to update the sample ID in the output type id but it was just to illustrate my comment on how I find it confusing.